### PR TITLE
Restrict permissions for the Windows agent in installation directory.

### PR DESCRIPTION
**Pull request by sergiovp**
|Related issue|
|---|
|#3520|

## Description

We discovered that in Windows, a non adminitrator user had read and execute permissions in `ossec-agent` directory.

We have to fix it on `.exe` and `.msi` files, we must disable ACL inheritance for the ossec-agent directory and copy the ACL from *Program Files*. Then, we remove the `grant X to Users` rule.

#### .EXE
We will apply this procedure in the setup-win.exe (post-install) tool:

1. Disable inheritance and copy ACL:
`ICACLS ossec-agent /inheritancelevel:d`

2. Remove permissions related to the "Users" group:
`ICACLS ossec-agent /remove "*S-1-5-32-545"`

#### .MSI
We will apply this procedure in the InstallerScripts.vbs file which is call in wazuh-installer.wxs file.

1. Create Windows Shell object:
`Set WshShell = CreateObject("WScript.Shell")`

2. Disable inheritance and copy ACL:
`setPermsInherit = "icacls """ & home_dir & "." & """ /inheritancelevel:d"`
`WshShell.run setPerms`

3. emove permissions related to the "Users" group:
`setUsersPerms = "icacls """ & home_dir & "." & """ /remove *S-1-5-32-545"`
`WshShell.run setPerms`

We did not fix the error on Windows XP operating system due to, in older versions than Windows Vista, we have to use `cacls` command instead of` icacls`. With `cacls`, we can not use * Well-known SIDs *, so we should refer to" Users "group in every language Windows XP is available.

### Implications:

+ The Administrator user will be allowed to access the agent folder and run the programs.
+ A user belonging to the Administrators group will be allowed to request permissions to access the agent.
+ An unprivileged user will be disallowed to access the agent unless it provides an administrator account credentials.

## Logs/Alerts example

As we can see, an unprivileged user will be disallowed to access the agent unless it provides an administrator account credentials.
image

## Tests

- Compilation without warnings in every supported platform
  - Linux  (not aplicable)
  - Windows Server 2019:
       - [x] .exe
       - [x] .msi
  - Windows Server 2016:
       - [x] .exe
       - [x] .msi
  - Windows Server 2012:
       - [x] .exe
       - [x] .msi
  - Windows 10 Home:
       - [x] .exe
       - [x] .msi
  - Windows Server 2008 x86 (32 bits)
      - [x] .exe
      - [x] .msi
  - MAC OS X (not aplicable)
- Source installation (not aplicable)
- [x] Package installation
- Source upgrade (not aplicable)
- [x] Package upgrade
- Memory tests
  - Valgrind report for affected components (not aplicable)
  - CPU impact (not aplicable)
  - RAM usage impact (not aplicable)
- Retrocompatibility with older Wazuh versions (not aplicable)
- Working on cluster environments (not aplicable)
- Configuration on demand reports new parameters (not aplicable)
- Review logs syntax and correct language (not aplicable)
- [x] QA templates contemplate the added capabilities https://github.com/wazuh/wazuh-qa/commit/2a8bc6cdb20ffb0c89891e4a488a0f210d7bce58

### CODE REVIEW

File: src/win32/wazuh-installer.wxs [hunk 1]
```diff
@@ -54,7 +54,7 @@
         <CustomAction Id="CustomAction_RemoveAllScript" BinaryKey="RemoveAllScript" VBScriptCall="removeAll" Return="check" Execute="commit" Impersonate="no" />
 
         <CustomAction Id="SetCustomActionDataValue" Return="check" Property="CustomAction_InstallerScripts" Value="&quot;[APPLICATIONFOLDER]&quot;,&quot;[ADDRESS]&quot;,&quot;[SERVER_PORT]&quot;,&quot;[PROTOCOL]&quot;,&quot;[NOTIFY_TIME]&quot;,&quot;[TIME_RECONNECT]&quot;" />
-        <CustomAction Id="CustomAction_InstallerScripts" BinaryKey="InstallerScripts" VBScriptCall="config" Return="check" Execute="commit" Impersonate="no" />
+        <CustomAction Id="CustomAction_InstallerScripts" BinaryKey="InstallerScripts" VBScriptCall="config" Return="ignore" Execute="commit" Impersonate="no" />
```
Comment (vikman90) [inline src/win32/wazuh-installer.wxs]: Why do we modify this behaviour?

Comment (bah07) [reply src/win32/wazuh-installer.wxs]: Finally, the old behavior has been maintained

**Comment by bah07** [comment]
Hi team,

- Permissions for ossec.conf have been removed for EveryOne group.
- Disabled inheritance in the ossec-agent root folder and remove permissions for the user group.

Regards.

**Review by bah07 — APPROVED** [review]
LGTM!

**Review by chemamartinez — APPROVED** [review]
LGTM

**Comment by vikman90** [comment]
@bah07 the MSI refuses to install on Windows XP SP3, while 4.0.0 does install.
image

May it be caused by the deletion of the call to `cacls`?

Best regards.

**Comment by sergiovp** [comment]
Hi,

It is true that `icacls` instruction is not available on Windows XP, you can check this [post](https://www.pcreview.co.uk/threads/is-the-command-icacls-exe-available-for-windows-xp-sp2.3234479/post-10976470).

It fails when `icacls` instruction is executed on `InstallerScripts.vbs` script.

You may use [cacls](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwi84fGJ--bsAhXFxYUKHbQ1AIcQFjAAegQIBRAC&url=https%3A%2F%2Fdocs.microsoft.com%2Fes-es%2Fwindows-server%2Fadministration%2Fwindows-commands%2Fcacls&usg=AOvVaw2boOrivxu_OBSnUUvV3NGI) or try with [xcacls](https://ss64.com/nt/xcacls.html) instructions.

Good luck and best regards.

**Comment by vikman90** [comment]
@sergiovp Thanks for the tip.

If I remember correctly, per our conversation, XP does not support user IDs, so we would need to use `cacls` and indicate the user group "Users" in the local system language.

That's why we decided to support permission restriction in Vista+ only. However, the MSI tries to run `icacls` on every platform. I think that is causing the problem.

Possible fix: let the MSI call `icacls` on Vista+ only.

Cheers!

**Comment by sergiovp** [comment]
Hi @vikman90 you are right.

It fails concretely on line 87 of `wazuh-installer.wxs` file:

~~~
<CustomAction Id="CustomAction_InstallerScripts" BinaryKey="InstallerScripts" VBScriptCall="config" Return="check" Execute="commit" Impersonate="no" />
~~~

`Return="check"` indicates that in case of error, the installation will be stopped. So Win XP tries to run `icacls` and gets an error  that causes the installation not to be completed.

You could change `Return="check"` to `Return=ignore` which means that the installation will continue in spite of `InstallerScripts.vbs` script returns an error, but as you say, I think the best option is to check the OS version before run `icacls` in `InstallerScripts.vbs` file  and in case of windows xp, not to run it.

I hope everything goes well. Best regards!

**Comment by vikman90** [comment]
Indeed, I've just added a guard in the VB script to prevent the installer from removing permissions (`icacls`) in XP and 2003.

So this is the final result:

|OS|Result|
|---|---|
|Windows XP SP3|This patch does not apply.|
|Windows Server 2003|This patch does not apply.|
|Windows Server 2016|OK.|
|Windows Server 2016|OK.|
|Windows Server 2019|OK.|
|Windows 10|OK.|

I think we're ready to merge this PR.

@sergiovp Thank you for your work on this issue.

I'm glad to read you around here. Best regards!

**Review by vikman90 — APPROVED** [review]

**Pull request merged**